api/src/main/java/org/openmrs/api/db/hibernate/HibernateCohortDAO.java

```diff
@@ -51,26 +53,34 @@ public void setSessionFactory(SessionFactory sessionFactory) {
 	public Cohort getCohort(Integer id) throws DAOException {
 		return (Cohort) sessionFactory.getCurrentSession().get(Cohort.class, id);
 	}
-	
+
 	/**
 	 * @see org.openmrs.api.db.CohortDAO#getCohortsContainingPatientId(java.lang.Integer)
 	 */
 	@SuppressWarnings("unchecked")
 	public List<Cohort> getCohortsContainingPatientId(Integer patientId) throws DAOException {
-		Query query = sessionFactory.getCurrentSession().createQuery(
-		    "from Cohort c where :patientId in elements(c.memberIds) and c.voided = false order by name");
-		query.setInteger("patientId", patientId);
-		return (List<Cohort>) query.list();
+		Disjunction orDisjunction = Restrictions.disjunction();
+        orDisjunction.add(Restrictions.isNull("endDate"));
+		orDisjunction.add(Restrictions.ge("endDate", new Date()));
+		return (List<Cohort>) sessionFactory.getCurrentSession().createCriteria(Cohort.class)
+                .add(Restrictions.eq("voided", false))
+                .createCriteria("members")
+                    .add(Restrictions.le("startDate", new Date()))
+                    .add(orDisjunction)
+                    .createAlias("patient", "p")
+                    .add(Restrictions.eq("p.patientId", patientId))
```

Comment:
You should add an argument for letting the caller to decided if you should included voided memberships and should default to false